### ISSUE TITLE
Autopopulate registries via registration functions

### ISSUE THREAD

**Issue by growlix** [mosaicml/composer#288]
## 🚀 Feature Request
Currently, all of the registries are [hard-coded in `trainer/trainer_hparams.py`](https://github.com/mosaicml/composer/blob/10564364cd6722a1e9a89efa1faa1868c948583a/composer/trainer/trainer_hparams.py#L39). It seems like it would be simpler to have a `register_{optimizer/model/etc.}` function that decorates the relevant object, as is currently being implemented for the generalized sample-prioritization algorithm that's superseding selective backprop (see the [registration function](https://github.com/mosaicml/composer/blob/matthew/sample_prioritization/composer/algorithms/selective_backprop/__init__.py#L25) and an [example use](https://github.com/mosaicml/composer/blob/matthew/sample_prioritization/composer/algorithms/selective_backprop/scoring_functions/irreducible_loss.py#L7)).

## Motivation
This would make it easier for users to add new/custom optimizers/models/etc. by reducing the number of places they have to make code changes.

**Comment by hanlint** [comment]
Yes! I believe we are moving towards this with  the auto-yahp work #269  (cc: @jbloxham ). The drawback of the decorater-as-registry approach is that the "menu" of options cannot be read from the code, but has to be a runtime query.

**Comment by hanlint** [comment]
Moving this to the YAHP repository.
